book/src/design/circuit/gadgets/sinsemilla.md

```diff
@@ -0,0 +1,160 @@
+# Sinsemilla
+
+## Overview
+Sinsemilla is a collision-resistant hash function and commitment scheme designed to be efficient in algebraic circuit models that support [lookups](https://zcash.github.io/halo2/design/proving-system/lookup.html), such as PLONK or Halo 2.
+
+The security properties of Sinsemilla are similar to Pedersen hashes; it is **not** designed to be used where a random oracle, PRF, or preimage-resistant hash is required. **The only claimed security property of the hash function is collision-resistance for fixed-length inputs.**
+
+Sinsemilla is roughly 4 times less efficient than the algebraic hashes Rescue and Poseidon inside a circuit, but around 19 times more efficient than Rescue outside a circuit. Unlike either of these hashes, the collision resistance property of Sinsemilla can be proven based on cryptographic assumptions that have been well-established for at least 20 years. Sinsemilla can also be used as a computationally binding and perfectly hiding commitment scheme.
+
+The general approach is to split the message into $k$-bit pieces, and for each piece, select from a table of $2^k$ bases in our cryptographic group. We combine the selected bases using a double-and-add algorithm. This ends up being provably as secure as a vector Pedersen hash, and makes advantageous use of the lookup facility supported by Halo 2.
+
+## Specification
+
+This section is an outline of how Sinsemilla works: for the normative specification, refer to [§5.4.1.9 Sinsemilla Hash Function](https://zips.z.cash/protocol/protocol.pdf#concretesinsemillahash) in the protocol spec.
+
+Let $\mathbb{G}$ be a cryptographic group of prime order $q$. We write $\mathbb{G}$ additively, with identity $\mathcal{O}$, and using $[m] P$ for scalar multiplication of $P$ by $m$.
+
+Let $k \geq 1$ be an integer chosen based on efficiency considerations (the table size will be $2^k$). Let $n$ be a **fixed** integer such that messages are $kn$ bits, where $2^n \leq \frac{q-1}{2}$. We use zero-padding to the next multiple of $k$ bits if necessary.
+
+$\textsf{Setup}$: Choose $Q$ and $P[0..2^k - 1]$ as $2^k + 1$ independent, verifiably random generators of $\mathbb{G}$, using a suitable hash into $\mathbb{G}$, such that none of $Q$ or $P[0..2^k - 1]$ are $\mathcal{O}$.
+
+$\textsf{Hash}(M)$:
+- Split $M$ into $n$ groups of $k$ bits. Interpret each group as a $k$-bit little-endian integer $m_i$.
+- $A_1 := Q$
+- for $i$ from $1$ up to $n$:
+  - $A_{i+1} := [2] A_i ⸭ P[m_i] = (A_i ⸭ P[m_i]) ⸭ A_i$
+- return $A_{n+1}$
```

Comment:
```suggestion
- let $A_0 := Q$
- for $i$ from $1$ up to $n$:
  - let $A_i := (A_{i-1} \;⸭\; S[m_i]) \;⸭\; A_{i-1}$
- return $A_n$

where ⸭ denotes incomplete point addition as defined in [Section 5.4.1.9 of the protocol specification](https://zips.z.cash/protocol/protocol.pdf#concretesinsemillahash).
```
The non-cosmetic changes here are:
* make the indexing match [Theorem 5.4.4 in the protocol spec](https://zips.z.cash/protocol/protocol.pdf#thmsinsemillaex);
* rename P to S to more closely follow the spec;
* delete [2] A_{i-1} ⸭ S[m_i], because this is *not* equivalent to (A_{i-1} ⸭ S[m_i]) ⸭ A_{i-1} in exceptional cases;
* define ⸭.